### PR TITLE
Migrate vending machine interfaces to TGUI

### PR DESCRIPTION
**Pull request by deadc4fe**
## What this does
Moves the vending machine interface to TGUI.
The wire UI has been changed from inline to the generic interface (as used for airlocks, etc).
This is mostly feature complete, the only missing feature being slogan editing (it wasn't fitting into the UI well, but I can add it if required), and the speaker switch has been moved to a wire, due to the aforementioned de-inlining of the wire UI.
### Before
before
old_edit_mode
old_confirm
### After
after
edit_mode
confirm
## Why it's good
Moves an old html based UI to a new pretty TGUI. One less spaghetti monster of HTML manipulation.

## Changelog
:cl:
 * tweak: Changed the UI for vending machines to TGUI.

### CODE REVIEW

**Comment by deadc4fe** [comment]
Oops, put the label in the wrong field.

**Comment by Eneocho** [comment]
This design made by an anon in the thread looks WAY better.
image

Also post a screenshot of the hacking menu.

**Comment by Cartlord** [comment]
is this more resource-hungry (for server and client) or less resource-hungry

**Comment by Eneocho** [comment]
TGUI tends to be slower on the client on load.

**Comment by deadc4fe** [comment]
While I agree the mockup does look nice, the category based backgrounds would require spriting and new variables somewhere, so I didn't go with that. A grid layout could look nice but most vendors have a lot less items than the MegaSeed Servitor. Regardless, if this is merged, updating the UI to something nicer down the line will be much easier.
> Also post a screenshot of the hacking menu.
hacking


> is this more resource-hungry (for server and client) or less resource-hungry


The javascript runs on the client (I assume), so it's slightly more client side than the original, which build UIs for clients by appending HTML blobs, but I assume it's probably negligable. There's also the overhead of TGUI, if we're counting that.

**Comment by KingMesh** [comment]
Don't do this, it both breaks the /vg/station artstyle and causes troubles for our Linux userbase.  If anything we should revert our TGUI more

**Comment by Kammerjunk** [comment]
soulless, unless,
browse() reappears

the change only
of TG >men

**Comment by tinysaturn** [comment]
vending machines is like the one thing that doesn't really need new UI

**Comment by gurfan** [comment]
> /vg/station artstyle

HTML UIs are clunky, slow, ugly, and spaghetti, the less HTML UIs the better.

**Comment by ConfusedCactus** [comment]
the /vg/station artstyle doesn't exist. if such a thing does exist then it would be best described as "codersprites" because it's chock full of them

File: code/game/machinery/vending.dm [hunk 1]
```diff
@@ -791,252 +836,7 @@ var/global/num_vending_terminals = 1
...
-		var/newslogan = input(usr,"Please enter a new slogan that is between 1 and [CUSTOM_VENDING_MAX_SLOGAN_LENGTH] characters long.","Add a New Slogan") as text
-		if(length(newslogan) > 0 && length(newslogan) <= CUSTOM_VENDING_MAX_SLOGAN_LENGTH)
-			product_slogans += html_encode(newslogan)
-
-	else if (href_list["delete_slogan_line"] && edit_mode && product_slogans.len > 0)
-		product_slogans -= product_slogans[text2num(href_list["delete_slogan_line"])]
-
-	src.add_fingerprint(usr)
-	src.updateUsrDialog()
+	tgui_interact(usr)
```
Comment (DamianX) [inline code/game/machinery/vending.dm]:
```suggestion
	tgui_interact(user)
```

File: code/game/machinery/vending.dm [hunk 2]
```diff
@@ -1228,7 +1028,7 @@ var/global/num_vending_terminals = 1
 			throw_item.throw_at(target, 16, 3)
 
 		src.visible_message("<span class='danger'>[src] launches [throw_item.name] at [target.name]!</span>")
-		src.updateUsrDialog()
+		tgui_interact(usr)
```
Comment (DamianX) [inline code/game/machinery/vending.dm]:
```suggestion
		SStgui.update_uis(src)
```

File: code/game/machinery/vending.dm [hunk 3]
```diff
@@ -766,6 +718,99 @@ var/global/num_vending_terminals = 1
...
+		data["contraband"] = extended_inventory
+		data["coin"] = coin != null
+		data["categories"] = list()
+		for (var/datum/data/vending_product/P in get_all_records())
+			var/list/product = list(
+				name = P.product_name,
+				amount = P.amount,
+				price = P.price,
+				category = P.category,
+				icon = P.mini_icon,
```
Comment (DamianX) [inline code/game/machinery/vending.dm]: Sending the base64 for all the icons with *every* interaction is the only thing in here that I find suspicious. How does a vending machine with a large inventory such as the Autodrobe perform with this?

**Comment by d3athrow** [comment]
Reminder calling something soulless is the nu-v equivalent of old thing good new thing bad.

Anyways you didn't take my thread suggestions

File: code/game/machinery/vending.dm [hunk 4]
```diff
@@ -369,12 +369,14 @@ var/global/num_vending_terminals = 1
...
 			amount = 1
 
+		var/obj/item/I = typepath
 		var/datum/data/vending_product/R = new()
 		R.product_path = typepath
 		R.amount = amount
 		R.original_amount = amount
 		R.price = price
 		R.assignedholiday = special
+		R.mini_icon = icon2base64(icon(initial(I.icon), initial(I.icon_state), SOUTH, 1))
```
Comment (d3athrow) [inline code/game/machinery/vending.dm]: Cache these please

**Comment by Kammerjunk** [comment]
calling something soulless is like calling it unfun

**Comment by zakanater19** [comment]
> the /vg/station artstyle doesn't exist.

what

**Comment by Eneocho** [comment]
> the /vg/station artstyle doesn't exist

Here we have an example on why you don't drink mercury, kids. Brain damage is a terrible curse to have.

**Comment by gurfan** [comment]
> > the /vg/station artstyle doesn't exist
> 
> Here we have an example on why you don't drink mercury, kids. Brain damage is a terrible curse to have.

retard

**Comment by WindowsErrors** [comment]
ooga booga change bad
thumps chest, grunting in caution

**Comment by zakanater19** [comment]
> > > the /vg/station artstyle doesn't exist
> > 
> > 
> > Here we have an example on why you don't drink mercury, kids. Brain damage is a terrible curse to have.
> 
> retard

nigger

**Comment by angryturnip** [comment]
Downvoting until you come up with a better design

**Comment by pedr69** [comment]
i like how the hacking menu now hides the products so you need to pay attention to the wires and the lights instead of just pulsing until the extra items show up

**Comment by ConfusedCactus** [comment]
> > the /vg/station artstyle doesn't exist
> 
> Here we have an example on why you don't drink mercury, kids. Brain damage is a terrible curse to have.

notice how he doesn't define the artstyle, which would be the obvious counterargument. why doesn't he do this? because there isn't one; there is no coherent theme to speak of.
certified eneocho moment

**Comment by zakanater19** [comment]
> > > the /vg/station artstyle doesn't exist
> > 
> > 
> > Here we have an example on why you don't drink mercury, kids. Brain damage is a terrible curse to have.
> 
> notice how he doesn't define the artstyle, which would be the obvious counterargument. why doesn't he do this? because there isn't one; there is no coherent theme to speak of. certified eneocho moment

join the game and take a screenshot and then compare to any other codebase

**Comment by Kammerjunk** [comment]
if you want specific criticisms, I think the dark-grey-and-darker-grey nannytrasen pattern adds a lot of ugly noise as a background to text. it makes the contents of the main body look too similar, and makes windows blend in rather than stand out.

beyond specific criticisms, it's just kinda bland, boring, uninspired. it looks like it was generated by an AI designed to create dark mode web designs without any context for content. it looks like it belongs in front 3/4 perspective walls and mothpeople. it reminds me of Hannibal Lecter's "you look like a rube" speech. it's not great.

**Comment by Hinaichigo** [comment]
If there are unsolved compatibility and performance issues with TGUI maybe it'd be good not to do this.
I like the icons, but I also think the design has too much negative space between categories.

Side note ー the other day I had an air alarm fail to load for me and I believe it was TGUI. And there have been other issues as well.

**Review by DamianX — COMMENTED** [review]
Code-wise this seems okay.
Have you tried the other themes?
I think it looks alright as it is, but several people said they dislike the grey colors.

**Comment by deadc4fe** [comment]
Right, the icon data has been moved to static data, as well as some other stuff that'll never change for a product, icons are now being cached and some redundant old stuff has been removed.


> i like how the hacking menu now hides the products so you need to pay attention to the wires and the lights instead of just pulsing until the extra items show up

You can have both windows open at once, I specifically put an update call on the cut and pulse procs for the vending wires.


> Code-wise this seems okay. Have you tried the other themes? I think it looks alright as it is, but several people said they dislike the grey colors.

I can spin up some of the other themes, though I'd rather not spam the PR, so I'll post those on the thread.

**Comment by Armadingus** [comment]
> Don't do this, it both breaks the /vg/station artstyle and causes troubles for our Linux userbase. If anything we should revert our TGUI more

Revert our new camera consoles with resizable windows and faster power monitors? You can't be serious.

**Comment by ghost** [comment]
yeah id like to see the mockup UI too, maybe with some keypad numbers you can click on like with real vendors, so entering "E6" gets the seed marked as E6

**Comment by deadc4fe** [comment]
I'm looking into doing the grid layout, but I'm not quite sure if there's any way to scale the icons up as nearest neighbour, which could be an issue if you want anything but a vending machine for ants.

**Comment by deadc4fe** [comment]
Ah, found it. Of course Microsoft has their own non-standard CSS rule for this. At least it works in IE.

**Comment by deadc4fe** [comment]
Well, I can't figure out a way to get something like that mockup actually looking nice, the main issue being that text scaling is a nightmare (scaling the grid in general (while looking nice), for that matter).
Here are the other themes for anyone who didn't see it in the thread, seeing as I'm too exhausted to mess around with webshit for another afternoon.
themes

**Comment by d3athrow** [comment]
@deadc4fe you can use zoom or scale probably along with setting -ms-interpolation-mode to nearest neighbor.

**Comment by BomberBro** [comment]
"To have lived through the transition stage of modern /vg/ makes a man feel preternaturally old; for here he is in modern times.... and yet he can himself distinctly remember the Middle Ages."
- Basil Hall Chamberlain, 8-year assistant main

There's nothing wrong with cleaning up the decrepit, blindingly white UIs that plague our ancient codebase. The "soul" that plagues the modern codebase of /vg/station, if there at all, comes from the souls of the damned that have perished trying to perform the Herculean feat of forming the kaleidoscope of different code and authors into something tangible. There's something to be said about keeping art styles somewhat in line with each other, but this UI change doesn't clash with them (given that the merch computer and others already use a similar style). 

The only real valid point I see is that this PR annihilates linux users. Do they deserve it? Yes. Should there be a toggle so that they can actually play the game? Also yes. If the toggle does get added, it should probably be in a new PR since this one has already been weighed down by the Will of the Masses™ and illiterate serfs like me are unlikely to check out an old PR again.  If the PR makes the current UI better, that's also great. Frankly, I'm just tired of having to scour the fetid jungles of the TheaterVend every time I want something.

**Comment by deadc4fe** [comment]
> @deadc4fe you can use zoom or scale probably along with setting -ms-interpolation-mode to nearest neighbor.

Yes, I did figure that out, the problem lies more in scaling the other elements, the product name text being the main problem. I'm not entirely sure trying new designs is the best use of my time at this point, though, given the current opinion on this PR. Messing around with the frontend isn't particularly fun to me anymore, that being said, I'm still for merging this PR, if we do actually want to move to TGUI in the end.

**Comment by Eneocho** [comment]
I like the malfunction and paper styles, add an clown eyerape style for the outfit vendor.

**Comment by KingMesh** [comment]
I like paper, though I think the buttons could be a more vibrant, stand out color than a pale yellow. Kind of like how the last vending machine had RBG text.

**Comment by djkramer123** [comment]
html menus blow, linux users deserve to suffer

**Comment by Eneocho** [comment]
Add a toggle that returns all menus (for the user only) to HTML, so linux players and players that dislike TGUI can still play.

**Comment by d3athrow** [comment]
we don't officially support playing the game on wine.

**Comment by angryturnip** [comment]
> we don't officially support playing the game on wine.

That's great, but I'm sure less people would downvote if there were a toggleable option for it

**Comment by bathosbathtime** [comment]
> Add a toggle that returns all menus (for the user only) to HTML, so linux players and players that dislike TGUI can still play.

I agree with your idea on principle, it is a net good for us to have more OSes be able to access the server, but I do have to ask who do you expect to do all of this work? I wouldn't know how to do it myself and this seems like quite a bit of work to be asking of someone else.

**Comment by Eneocho** [comment]
> > Add a toggle that returns all menus (for the user only) to HTML, so linux players and players that dislike TGUI can still play.
> 
> I agree with your idea on principle, it is a net good for us to have more OSes be able to access the server, but I do have to ask who do you expect to do all of this work? I wouldn't know how to do it myself and this seems like quite a bit of work to be asking of someone else.

I use win 10 and TGUI runs like ass for me, taking 2-5 seconds to load the interface each time, I just want the option of having the ugly but functional interface instead of the pretty but awkward and slow one.
If the guy wants to add TGUI to one of the most used machines, then he also should add such a toggle for the very clear majority that does not want it (at least for the vending machine UI). Is it really that hard to add a toggle between two UI styles? We already have current html UI and this TGUI coded (for the most part).

**Comment by DamianX** [comment]
"Just do twice the work now and going forward" is a very non-coder opinion to have, there's no way to put this that doesn't make me sound like an ass.
We don't have coders to maintain one UI, let alone two.
Keeping both is just not a sensible option.

**Comment by Eneocho** [comment]
Fair, then keep the shittier UI that doesn't lag instead of the fancy laggy one.

**Comment by d3athrow** [comment]
`I use win 10 and TGUI runs like ass for me, taking 2-5 seconds to load the interface each time`

Sounds like an issue to be resolved, it should take a bit of time the first load then be instant the next.

**Comment by gurfan** [comment]
> I use win 10 and TGUI runs like ass for me, taking 2-5 seconds to load the interface each time

Which interface in particular?

**Comment by ConfusedCactus** [comment]
> > > > the /vg/station artstyle doesn't exist
> > > 
> > > 
> > > Here we have an example on why you don't drink mercury, kids. Brain damage is a terrible curse to have.
> > 
> > 
> > notice how he doesn't define the artstyle, which would be the obvious counterargument. why doesn't he do this? because there isn't one; there is no coherent theme to speak of. certified eneocho moment
> 
> join the game and take a screenshot and then compare to any other codebase

image

define the coherent theme linking together all elements of this artstyle for me, if you would
surely you can do this since you have such conviction that /vg/ has a coherent artstyle

**Comment by Hinaichigo** [comment]
I'm still not convinced that the hit to performance and usability is worth the aesthetic change. And like I said the new screens have less information density on them. Too much prioritizing of form over function. Keeping things lean and smooth is important to avoid things getting bloated, buggy, and laggy like other servers. It might be better to hold off for now and find a better solution that both looks nice and performs well.

**Comment by gurfan** [comment]
> I'm still not convinced that the hit to performance and usability is worth the aesthetic change. And like I said the new screens have less information density on them. Too much prioritizing of form over function. Keeping things lean and smooth is important to avoid things getting bloated, buggy, and laggy like other servers. It might be better to hold off for now and find a better solution that both looks nice and performs well.
> 
> I also have occasional issues with TGUI interfaces not loading properly, using W10, while the current html ones function without a hitch.

Again, which interface in particular?

**Comment by Hinaichigo** [comment]
~~Most recently it was the air alarm.~~

~~and someone else was having issues with the isolation centrifuge.~~

**Comment by gurfan** [comment]
> Most recently it was the air alarm.
> 
> and someone else was having issues with the isolation centrifuge.

Neither of those use TGUI, they use NanoUI.
The only things using TGUI right now IIRC are the Camera Computer, GPS, Traitor Uplink, Power Computer, Mecha/Cyborg Computers, TEG, Merch Computer

I've found the things that use TGUI to be faster and more reliable but that's mostly just anecdotal evidence; I realize it won't apply to everyone. Overall I think the hit to performance is being overstated.

**Comment by Viciousaurus** [comment]
> > > > > the /vg/station artstyle doesn't exist
> > > > 
> > > > 
> > > > Here we have an example on why you don't drink mercury, kids. Brain damage is a terrible curse to have.
> > > 
> > > 
> > > notice how he doesn't define the artstyle, which would be the obvious counterargument. why doesn't he do this? because there isn't one; there is no coherent theme to speak of. certified eneocho moment
> > 
> > 
> > join the game and take a screenshot and then compare to any other codebase
> 
> image
> 
> define the coherent theme linking together all elements of this artstyle for me, if you would surely you can do this since you have such conviction that /vg/ has a coherent artstyle

SOVL, clearly

**Comment by Eneocho** [comment]
> > I use win 10 and TGUI runs like ass for me, taking 2-5 seconds to load the interface each time
> 
> Which interface in particular?

Merch vendor and the power monitor also the few times I've used the sec cameras. And whenever I open the uplink on the offchance I roll tator.

**Comment by Eneocho** [comment]
The other TGUIs run somewhat smoothly as far as I can remember, but those 4 run like ass each time.

**Comment by Exxion** [comment]
>we don't officially support playing the game on wine.

Indeed, but it's also moot because TGUI works on Wine anyway

**Comment by WindowsErrors** [comment]
I would not say the hit to performance is overstated, I think it's stated just right

how do I just turn off tg ui

**Comment by Eneocho** [comment]
> 
> how do I just turn off tg ui

No such feature and as damian implied it is not likely to be a feature anytime soon.

**Comment by optimumtact** [comment]
hahah you're so ass backwards

**Comment by cloutcode** [comment]
go add some pride flag jumpsuits demon; leave this place

**Comment by ExpulsionSnooper** [comment]
ENTER

**Comment by bathosbathtime** [comment]
I have made a server poll for this PR. It will finish in a week at 05/21/2022

**Comment by WindowsErrors** [comment]
The Space Chimpanzee Wars of BYOND National Park had been raging and ravaging for 12 years, leaving thousands of hurt feelings and community abandonments on all sides

In the comments sections and forums, people have been asking the question: what is the price of Soul?
How many more must be angered?

9 years after the Settlement of the Thread, dead cfourfe's first pull request is submitted
For the committal of this historical code, the software developer of GitHub is chosen

While the others were busy huddled around their glowing screens clicking thumb icons, the thumb downers used the opportunity to call attention to the pull request
However these two other monkey brothers Cactus and gurfan didn't want to give up the thumbing up of the pull request so easily, but they knew neither of them would be able to beat them in a thumb war alone, so they put their monkey brains together in an interaction that probably went like:

"What if instead of trying to beat thumb down in the vote one thumb on one thumb, we gang up on down and take them out together with our two thumbs?"
"Ok"

From this, 16 males and 1 female clicked thumbs pointed up to create the thumb uppers
Thumb downers didn't really like that and kind of went bananas
You could say they chimped out, turned into Furious George

Basically, Ennyokoh and two dozen of his best monkey buds gathered together to create a war party and ambush dj kramer, a thumb upper male while he was just on his own minding his business
They just beat his groin with wet pool noodles and when he couldn't move anymore they celebrated by jumping up and down and hollering like a bunch of baboons

Over the next few days they used some gorilla warfare tactics and took out the thumb uppers one by one like this

Eventually the next one on the list was west even though he was super chill whenever the thumb downers came down and ganked one of his brothers in front of him
The cure to the Chinese flu still hasn't quite been perfected yet so he ended up being physically handicapped and wasn't much trouble for the thumb downers considering how he couldn't even stand up on two legs to begin with

Last on the list was Kammer, an innocent little chimp who just wanted to get away with her thumbs intact
The shrimpanzee managed to avoid the fighting with the thumb downers and hide out for a whole week before subsequently reappearing
She was actually teaching herself to use an old M16 that one of the guards left unsecured at the research compound and ended up filling all but three of the thumb downers with enough lead to give an elephant vomiting, severe weight loss, and regular seizures before the other thumb downers could deal with her

Now that all the thumb uppers were gone, the remaining thumb downers that were still alive decided to forever ignore the old pull request for the rest of time now that they have fulfilled their contribution to the Will of the Masses™ 

Order and peace were restored with the unwanted being eliminated from the community
Some of the players return to the game
Others remain roaming the GitHub a bit longer because they find the rest of the community fun again
Culture, arts, and science go on to flourish in a golden age

___

A minor skirmish in a war with no end...

**Comment by Exxion** [comment]
> the /vg/station artstyle doesn't exist. if such a thing does exist then it would be best described as "codersprites" because it's chock full of them

I know I'm almost a week late to this discussion but the responses to this post are very funny
This statement was, to my memory, pretty much totally uncontroversial a few short years ago, in which time we haven't made any real fundamental changes to the art
Our art only seems more consistent now than it did then because all the *other* codebases have changed so much in appearance that our style gets to be united as "The Old One"
To be clear, I'm not saying that's a bad thing. I think most of the artstyle overhauls elsewhere have generally made everything look worse. It's just funny to see these reactions that sound like people can't even imagine someone having this position that was near-universal not that long ago.

**Comment by Kurfursten** [comment]
@Exxion When I read your comment, it sounded familiar, and I wondered, is it possible that Exxion has said almost this exact same thing half a decade ago? And after a little digging I found it.

https://github.com/vgstation-coders/vgstation13/pull/13934#issuecomment-279152051

We are becoming old men

**Comment by bathosbathtime** [comment]
wow those air tank sprites look awful

**Comment by WindowsErrors** [comment]
if the /vg/ art style is HTML codersprites then the TG art style is glossing everything over and weirdly canting it at a 45° angle so the studio lights on the station's ceiling reflect back directly into the camera lens

they're both valid art styles, I just like the first one more

**Comment by SonixApache** [comment]
use the new one but make it look like the old one
functional + soul

**Comment by Kurfursten** [comment]
Looks like based on the poll, the support isn't there. That being said, it looks like you put a lot of work into this, so I think if you took feedback and tried again in another PR you might be successful. Thanks for your contribution.

**Comment by Optimism333** [comment]
sad

**Comment by WindowsErrors** [comment]
bro [even space parallax has a toggle](https://github.com/vgstation-coders/vgstation13/pull/10085) meanwhile the people here are wondering why we're resisting tgui so much

**Comment by wotisjanitor** [comment]
I came here to downvote this because its not vgui

**Comment by Exxion** [comment]
image

**Pull request closed without merging**